staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt_test.go

```diff
@@ -0,0 +1,409 @@
+		// NOTE: C >= serverConcurrencyLimit or else no packets are sent via Work
+		C: 100,
+	}
+	// Run blocks until all work is done.
+	w.Run()
+
+	if count != int64(requests) {
+		t.Errorf("Expected to send %d requests, found %v", requests, count)
+	}
+}
+
+// TestReqMgmtMultiple verifies that fairness is preserved across 5 flowschemas
+// and that the serverConcurrencyLimit/assuredConcurrencyShares are correctly
+// distributed
+func TestReqMgmtMultiple(t *testing.T) {
+	// init objects
+	groups := 5
+	flowSchemas := []*rmtypesv1a1.FlowSchema{}
+	for i := 0; i < groups; i++ {
+		flowSchemas = append(flowSchemas,
+			&rmtypesv1a1.FlowSchema{
+				ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("fs-%d", i)},
+				Spec: rmtypesv1a1.FlowSchemaSpec{
+					PriorityLevelConfiguration: rmtypesv1a1.PriorityLevelConfigurationReference{
+						fmt.Sprintf("plc-%d", i),
+					},
+					MatchingPrecedence: int32(i), // TODO(aaron-prindle) currently ignored
+					// DistinguisherMethod: *rmtypesv1a1.FlowDistinguisherMethodByUserType, // TODO(aaron-prindle) currently ignored
+					// Rules: []rmtypesv1a1.PolicyRuleWithSubjects{}, // TODO(aaron-prindle) currently ignored
+				},
+				// Status: rmtypesv1a1.FlowSchemaStatus{},
+			},
+		)
+	}
+	priorityLevelConfigurations := []*rmtypesv1a1.PriorityLevelConfiguration{}
+
+	for i := 0; i < groups; i++ {
+		priorityLevelConfigurations = append(priorityLevelConfigurations,
+			&rmtypesv1a1.PriorityLevelConfiguration{
+				ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("plc-%d", i)},
+				Spec: rmtypesv1a1.PriorityLevelConfigurationSpec{
+					AssuredConcurrencyShares: 1,
+					Exempt:                   false,
+					GlobalDefault:            false,
+					HandSize:                 int32(8),
+					QueueLengthLimit:         int32(65536),
+					Queues:                   int32(128),
+				},
+			},
+		)
+	}
+
+	serverConcurrencyLimit := groups
+	requestWaitLimit := 5 * time.Second
+	requests := 100
+
+	countnum := len(flowSchemas)
+	counts := make([]int64, len(flowSchemas))
+
+	delegate := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(500 * time.Millisecond)
+		for i := 0; i < countnum; i++ {
+			if r.Header.Get("PRIORITY") == strconv.Itoa(i) {
+				atomic.AddInt64(&counts[i], int64(1))
+			}
+		}
+	})
+
+	server, _ := createRequestManagementServerAndClient(
+		t,
+		delegate, flowSchemas, priorityLevelConfigurations,
+		serverConcurrencyLimit, requestWaitLimit,
+	)
+	// defer server.Close()
+
+	ws := []*Work{}
+	for i := 0; i < countnum; i++ {
+		header := make(http.Header)
+		header.Add("PRIORITY", strconv.Itoa(i))
+		req, _ := http.NewRequest("GET", server.URL, nil)
+		req.Header = header
+
+		ws = append(ws, &Work{
+			Request: req,
+			N:       requests,
+			// NOTE: C >= serverConcurrencyLimit or else no packets are sent via Work
+			C: serverConcurrencyLimit,
+		})
+
+	}
+	for _, w := range ws {
+		w := w
+		go func() { w.Run() }()
+	}
+
+	time.Sleep(1 * time.Second)
```

Comment:
Isn't 1 second too short to serve 100 requests that each takes half a second?

Reply:
The idea of the test is that we send a "high" load of requests to the handler and then cut off before all requests are served.  When we cut off, we should see that the requests were served fairly (same # of requests served for each flowschema).  With the requests taking some time it should be easier to cut off at the point at which they are even.  In the case that the test waited for all of the requests to be served, the test would only be testing that all the requests got served and not the fairness.